src/alacris_lib/operator_contract.ml

```diff
@@ -24,58 +27,166 @@ let topic_of_hash (hash : Digest.t) : Bytes.t option =
   Some (encode_function_parameters [abi_digest hash])
 
 
-let retrieve_contract_address_quadruple : Digest.t -> (Address.t * Digest.t * Digest.t * Revision.t) Lwt_exn.t =
-  fun transaction_hash ->
-  let open Lwt_exn in
-  Ethereum_json_rpc.eth_get_transaction_receipt transaction_hash
-  >>= function
-  | None -> bork "No tx receipt for contract creation"
-  | Some receipt ->
-     let contract_address = Option.get receipt.contract_address in
-     let contract_block_number = receipt.block_number in
-     Ethereum_json_rpc.eth_get_transaction_by_hash transaction_hash
-     >>= fun trans ->
-     let code_hash = Digesting.digest_of_string (Bytes.to_string trans.input) in
-     return (contract_address, code_hash, transaction_hash, contract_block_number)
-
-
-let operator_contract_log = false
-
 
-let contract_address = ref Address.zero
-let contract_block_number = ref Revision.zero
 
-let set_contract_address address = contract_address := address
-
-let set_contract_block_number blk_number = contract_block_number := blk_number
-
-let get_contract_address () =
-  if operator_contract_log then
-    Logging.log "get_contract_address : contract_address=%s" (Address.to_0x !contract_address);
-  !contract_address
-
-let get_contract_block_number () =
-  if operator_contract_log then
-    Logging.log "get_contract_block_number : contract_block_number=%s" (Revision.to_string !contract_block_number);
-  !contract_block_number
 
 (** build the encoding of a call to the "deposit" function of the operator contract
     address argument is the operator *)
 let make_deposit_call : operator:Address.t -> contract_address:Address.t -> Ethereum_chain.Operation.t =
   fun ~operator ~contract_address ->
+  if operator_contract_log then
+    Logging.log "Beginning of make_deposit_call";
   let parameters = [ abi_address operator ] in
   let call = encode_function_call { function_name = "deposit"; parameters } in
   Operation.CallFunction (contract_address, call)
 
+
+
 let pre_deposit : operator:Address.t -> amount:TokenAmount.t -> contract_address:Address.t -> PreTransaction.t =
   fun ~operator ~amount ~contract_address ->
+  if operator_contract_log then
+    Logging.log "Beginning of pre_deposit";
   let oper = make_deposit_call ~operator ~contract_address in
   PreTransaction.{operation=oper; value=amount; gas_limit=Side_chain_server_config.deposit_gas_limit}
 
 
 
+
+type contract_address_config_input =
+  { contract_address : string
+  ; code_hash : string
+  ; creation_hash : string
+  ; creation_block : int
+  } [@@deriving of_yojson]
+
+
+type contract_address_config =
+  { contract_address : Address.t
+  ; code_hash : Digest.t
+  ; creation_hash : Digest.t
+  ; creation_block : Revision.t
+  }
```

Comment:
Shouldn't the two be the very same type, using the Address and Digest and Revision yojson support to read the string properly?

Reply:
This is for sure the natural solution.
However, the problem is that the Digest.t does not have a "of_yojson" implementation.
I tried to implement the Digest.of_yojson but without success.